### ISSUE TITLE
Message handler invoked multiple times despite using outbox

### ISSUE THREAD

**Issue by pardahlman** [Particular/NServiceBus#5648]
We are using NServiceBus (7.2.4) with RabbitMQ (5.2.0) as the underlying transport. Some of our message handlers performs operations that _has side effects_ and _can't be enlisted in a transaction_. We rely on the outbox feature for exactly once delivery to prevent unwanted processing of message duplicates.

Despite this, there are scenarios where the message handler is fully executed with a message duplicate. I'm seeing this when the client (NServiceBus endpoint) loses its connection to the RabbitMQ broker. When this happens, the client is trying to reconnect and once connection has been reestablished it processes the message. At the same, the un-acked message is returned to the broker as the connection is lost. It is then delivered to an instance of the application still connected to the broker and is handled there.

Only one of the message handler transaction will be able to complete, as there will be a primary key violation [when the outgoing operations are stored in the outbox](https://github.com/Particular/NServiceBus/blob/851a5e91320798cb2101c88f53092a00594e7c20/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs#L40). However, the operation with side effect will still have executed twice. 

It seams to me like the [outbox/deduplication code](https://github.com/Particular/NServiceBus/blob/851a5e91320798cb2101c88f53092a00594e7c20/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs#L24-L60) may invoke a message handler with a message duplicate if it or some other instance is in the process of handling the "original" message. Perhaps this could be avoided if the [deduplication check](https://github.com/Particular/NServiceBus/blob/851a5e91320798cb2101c88f53092a00594e7c20/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs#L27) also wrote to the database that it was in the process of handling a message, and that any other message received with the same ID would be considered a duplicate if there was a record of it being handled?

**Comment by danielmarbach** [comment]
@pardahlman What you are seeing happening in the case with rabbitmq this is by design of the current outbox. The outbox transaction spawns the message processing pipeline and writes the outbox message after the execution of the handling pipeline. If during the execution (not yet stored) the connection dies or in cases of transport that use visibility timeouts the message visibility timeout is due and the message reappears it will not find a de-duplication entry and therefore execute the handling pipeline again like you said. 

The general recommendation for side effects like RPC calls that cannot enlist in a transaction is to for example use a message that you `SendLocal` to yourself. Then this problem is gone. Should more state management be needed then a saga could manage that and also use SendLocals to dispatch the side effects to local handlers. 

About the solution you proposed. Thanks for brainstorming with us! As far as I remember what you are talking about comes close to a pessimistic lock approach that would require a lease time to avoid deadlocks. That would be a different design approach to the current outbox design. We are working on various persisters on the possibility to opt-in for pessimistic locking for sagas. So if a saga persister supports that you could potentially move the side effect into a saga under a pessimistic lock which would drastically reduce the possibility of having the side effect occurring multiple times. The problem though is that having such an effect in a saga can increase the contention on the saga and that's why our recommendation generally is to also use for example SendLocal for those side effects in the saga.

**Comment by pardahlman** [comment]
Can you elaborate on how the use of `SendLocal` would eliminate this problem? Isn't it the case that the handler for the local command can lose connection to the broker _after_ the RPC call is made but _before_ it is stored in the outbox? In this case, the message will be returned to the broker and delivered to another instance, since `SendLocal` send to any instance of the logical endpoint rather than a specific one.

I'm probably missing something, would be great if there is a deterministic solution using `SendLocal` 👌

**Comment by danielmarbach** [comment]
Correct the problem can still occur like you mentioned. There is no 100% bullet proof situation that solves it. This solution makes it just slightly less likely to occur because the operations are decoupled over the broker. The assumption is though that due to isolation the side effect occurring thing from the other operations that can enlist in the transaction you have smaller execution and failure units that are decoupled from each other.

I mean even with a regular handler if you have a business logic flaw that throws an exception the side effect could occur multiple times due to the recoverability kicking in and doing immediate and second level retries. At the end of the day if you need more reliability you need to make sure that the broker is actually highly available to fulfill your SLAs. But even then you might not have entirely eliminated the problem because fallacies of distributed computing. 

Can you maybe talk a bit more about this resource that you are calling? What are the business consequences of in the worst case calling it multiple times? What would be the consequences of not calling it?

**Comment by pardahlman** [comment]
I see, thank you for clarifying 🙏 .

We have some specific places where it very much undesired to run the handler code more than once, but we also have a large code base and want to be able to give some assurances as to what the messaging infrastructure guarantees. With regard to that, we end up in a bit of a strange position because we have not been able to enlist our domain db-calls etc in the outbox transaction (mostly because a lot of our code relies on DI, but the outbox transaction is passed through the `IMessageHandlerContext` and it would be a massive job to rewrite all this code -- and conceptually we are not sure it is desired to entangle the messaging framework with the database layer). Instead we used the outbox with an inner UnitOfWork/`TransactionScope`. However, when the reconnect scenario above happens, the inner UnitOfWork will complete before the outbox transaction fails and the operation has partially completed.

In short my understanding of the guarantees we can make given this setup has gone from

* _`at-least-once` message delivery_ to
* _perhaps delivered more than once, but only completed once_ and finally to 
* _delivered multiple times and partially committed_.

Of course, the situation would improve if it was possible to configure the outbox to use a transaction scope, but I believe there is no such option?

**Comment by danielmarbach** [comment]
> With regard to that, we end up in a bit of a strange position because we have not been able to enlist our domain db-calls etc in the outbox transaction (mostly because a lot of our code relies on DI, but the outbox transaction is passed through the `IMessageHandlerContext` and it would be a massive job to rewrite all this code

yeah understandable. The design of the handler context is our not so gentle push to make sure customers are not intertangling their business code with messaging stuff and leave those side effects in the handlers. At the time when we designed it we felt this is especially crucial due to the async nature of those APIs so that in case a customer has a big code base that doesn't use async/await yet the viral nature of those APIs could be contained mostly within the handler code. I'm aware though that would require you to also move your DB stuff into the handler and that is a too big undertaking.

Have you thought about the following approaches?

## Method injection of connection and transaction

When you call your code that would ideally want to enlist into the outbox transaction you pass the adapted connection and transaction into the "lower layers" like 

```
public async Task Handle(Message message, IMessageHandlerContext context) {
   await lowerLayers.DoYourMagic(context.SynchronizedStorage.Get...);
}
```

and then pass it through. This would make it apparent that your code relies on an ambient connection and transaction and potentially be more "intention revealing"

## Prepare the ambient connection and transaction for DI from the pipeline

The gist would be that you register a component on the container as a singleton that looks something like this

```
public class AmbientConnectionAndTransactionProvider
{
    AsyncLocal<DbConnection> connectionHolder = new AsyncLocal<DbConnection>();
    AsyncLocal<DbTransaction> transactionHolder = new AsyncLocal<DbTransaction>();

    public DbConnection Connection
    {
        get => connectionHolder.Value;
        set => connectionHolder.Value = value;
    }

    public DbTransaction Transaction
    {
        get => transactionHolder.Value;
        set => transactionHolder.Value = value;
    }
}
```

The behavior that sets the ambient context

```
class Behavior : Behavior<IInvokeHandlerContext>
{
    public override Task Invoke(IInvokeHandlerContext context, Func<Task> next)
    {
        var session = context.SynchronizedStorageSession.SqlPersistenceSession();
        var provider = context.Builder.Build<AmbientConnectionAndTransactionProvider>();
        provider.Connection = session.Connection;
        provider.Transaction = session.Transaction;

        return next();
    }
}
```

a sample registration using the confirming container for simplicity

```
public static class EndpointConfigurationExtensions
{
    public static void RegisterMagic(this EndpointConfiguration configuration)
    {
        configuration.Pipeline.Register(new Behavior(), "Extracts the synchronized storage and makes it available for injection");
        configuration.RegisterComponents(c =>
        {
            // just for the demo
            c.ConfigureComponent<MyComponent>(DependencyLifecycle.InstancePerUnitOfWork);

            c.RegisterSingleton(new AmbientConnectionAndTransactionProvider());
        });
    }
}
```

the handler and the component

```
public class ShipOrderHandler :
    IHandleMessages<ShipOrder>
{
    static ILog log = LogManager.GetLogger<ShipOrderHandler>();
    MyComponent component;

    public ShipOrderHandler(MyComponent component)
    {
        this.component = component;
    }

    public Task Handle(ShipOrder message, IMessageHandlerContext context)
    {
        log.Info($"Order Shipped. OrderId {message.OrderId}");
        return component.Do();
    }
}

public class MyComponent
{
    AmbientConnectionAndTransactionProvider provider;

    public MyComponent(AmbientConnectionAndTransactionProvider provider)
    {
        this.provider = provider;
    }

    public Task Do()
    {
        // we can only resolve here
        var connection = provider.Connection;
        var transaction = provider.Transaction;

        Debug.Assert(connection.State == ConnectionState.Open, "connection must be open");
        Debug.Assert(transaction.Connection == connection, "transaction must be managed by connection");
        return Task.CompletedTask;
    }
}
```

As you can see the draw back is though that the connection and the transaction method are only available when the handler calls the dependency and not yet at constructor time. So the temporal coupling is suddenly exposed. Some people like that some don't. If you want to hide it away you would either need to resolve `MyComponent` in the above example in the handler and then call `Do` (my preference) or decorate `DbConnection` as well as `DbTransaction` to delegate the calls to the underlying real lazy connection or transaction. With that you should be able to enlist into the outbox transaction without needing to rewrite a lot of code. 

What do you think?

**Comment by pardahlman** [comment]
Thanks for the detailed response, much appreciated!

We have considered both approaches, but had to dismiss the explicit passing of of synchronized storage arguments to database layer since it would change the APIs in some of our most used services that is called in thousands of places. I do think it is a clean solution, but unfortunately not feasible due to the extra work it would entail.

We have also explored `AsyncLocal` and actually use it in another NServiceBus/DI related situation (this is related to design decisions we have made, outside the scope of this discussion). While it works fine, it adds an unexpected complexity and (IMHO) feels hacky.

I see two ways that I think would work fine in our code base:

1. As I eluded in my previous comment, if there where an option for the outbox to enlist in a ambient `TransactionScope` that our database code could seamlessly enlist to
2. Register a scoped per-message service in the DI that could be resolved in the dependency graph for the handler.

That being said, I understand and appreciate that there are a lot of other factors to consider if such impactful changes would be implemented, but it would indeed be something on my wishlist 😉

**Comment by pardahlman** [comment]
Yes, sorry for being vague - I meant just that: a DI-resolvable service that can provide the current `SynchronizedStorageSession` like you described would more than suffice. I wouldn't want to implement it myself with `AsyncLocal`, but if it came with NSB and was vetted, battle tested and maintained by Particular I wouldn't mind 😸

**Comment by danielmarbach** [comment]
@pardahlman yeah I get that. I tracked this feature request also internally to make sure it is getting triaged.

One last thing that I think would be a reasonable trade off that doesn't require you to change much of your code plus would get rid of the hackyness and complexity of what we discussed. What about if you change your handlers from

```
public class MyHandler {
   public MyHandler(MyDependencyThatRequiresSomewhereConnectionAndTransaction dependency) {}

   public async Task Handle(...) {
      await dependency.Do
   }
}
```

to

```
public class MyHandler {
   public MyHandler(FactoryForMyDependencyThatRequiresSomewhereConnectionAndTransaction factory) {}

   public async Task Handle(...) {
      
      var session = context.SynchronizedStorageSession.SqlPersistenceSession();
      var dependency = factory.Create(session.Connection, session.Transaction);
      await dependency.Do();
   }
}
```

with that all the underlying code that is in the hierarchy of `MyDependencyThatRequiresSomewhereConnectionAndTransaction` doesn't have to change and for you handlers the change is really a minor one because instead of constructor injection you switch to factory injection.

**Comment by pardahlman** [comment]
Thank you again for helping out, @danielmarbach -- it is very much appreciated 🙏 

Many message handlers in our code base have the domain logic implemented in the handler itself, so the layer "`dependency`" seldom exists. Actually, we think this works fine, since we often just store something in the database, publish an event and that's it! In these cases an additional layer would just bring one more level of indirection to navigate through.

But for the sake of this argument, say that we would have this layer everywhere - then we would need to update our message handlers to do something similar to your code (perhaps with a `IServiceProvider` and `ActivatorUtilities.CreateInstance` since we're using Microsoft DI). This would be a significant effort, and come with some significant drawbacks - from the top of my mind:

* The checks that NSB does that `IEndpointInstance` is not injected in a handler would be bypassed, because such dependency would be resolved in the handler rather than by the NSB-builder
* The ServiceCollection's validate dependency graph on build would fail as some of the dependencies would not be registered

To me, the code that you suggest we put in each of our handlers looks like something that would fit in the NSB framework, except that it would resolve a handler instead of the "`depdendency`" layer. 😛 

Again, I want to say thank you for taking the time to help out and discuss this 👌

**Comment by pardahlman** [comment]
My apologies for not getting back to you earlier.

I don't think I'm at a liberty to share code, unfortunately. There has been a development at out side, however: we had an discussion internally and decided that the `AsyncLocal` approach might be acceptable for us until there are any better alternatives available. This will prevent the worst outcome (_delivered multiple times and partially committed_) and "only" leave us at _delivered multiple times with only one successful commit_.

As I'm wrapping up this issue I just want to re-iterate some of the points made:

* It would be most desired if an ongoing message process would prevent a duplicate to start being processed. (Alternatively, change behavior in the RabbitMQ transport so that messages processing would be aborted when connection to the broker is lost)
* A welcome feature would be if the outbox transaction could be configured to use/be/enlist in a ambient transaction scope. (This would also remove the "hidden incompatibility" with UnitOfWork)
* An DI injectable service that provides the current message process' transaction and connection would also be a nice addition.

Thank you for helping out and listening to our problems and my rants 😉 ! Looking forward to track any issue that relates to any of these points. Feel free to close this issue, or reply if you want to keep discussing any part of it!

**Comment by SzymonPobiega** [comment]
Hi @pardahlman 

We are currently in the process of preparing an enhancement release of SQL-related technologies. Some of the things we listed are on out TODO list (e.g. [TransactionScope in the Outbox](https://github.com/Particular/NServiceBus.NHibernate/issues/274), [outbox pessimistic mode](https://github.com/Particular/NServiceBus.Persistence.Sql/issues/37) or [injecting Synchronized Session via DI container](https://github.com/Particular/NServiceBus.Persistence.Sql/issues/375)) but we can't yet promise which issues will eventually be included in the release.

**Comment by pardahlman** [comment]
TransactionScope in the outbox and outbox pessimistic mode would solve ALL of our problems, given that it would be implemented for MSSQL (and not just NHiberate).

It is a bit disconcerting that these issues are over three years old. Is it possible to get any indication if they are actively worked at or just "idea placeholders"? The latter one was marked as `candidate-for-next-release` 2018, so I assume that it is not entirely up-to-date?

Thanks!

**Comment by SzymonPobiega** [comment]
Hi

We have just released V6 of SQL Persistence which registers `ISqlStorageSession` in the DI infrastructure so you can inject it into your repositories.

**Comment by pardahlman** [comment]
This is very good news, @SzymonPobiega - thank you for the hard work and all effort put into this 🙏 . We will be looking to upgrade shortly to this release!

**Comment by SzymonPobiega** [comment]
Since the change has been released I am closing this one now. @pardahlman feel free to re-open it or create a new one if you find the latest changes not good enough to handle your scenario.
